### DISCUSSION TITLE
Generic keys for Redis Proxy Managers

### DISCUSSION THREAD

**Discussion by namhptran** [vladimir-bukhtoyarov/bucket4j#358] [Ideas]
Hi @vladimir-bukhtoyarov 

Saw https://github.com/bucket4j/bucket4j/issues/351 and thought that it would be nice to have a similar change done for Redis-backed Proxy Managers.

I use a LettuceBasedProxyManager but my keys are usually Strings, meaning I always have to use `String#getBytes(StandardCharsets.UTF_8)` to convert my Strings into `byte[]` before using as a key. This also means hotswapping ProxyManagers between different implementations for different configurations using Spring dependency injection isn't always so simple.

Is this a feature on the radar for anytime soon?

For Lettuce at least, they provide [RedisCodec](https://lettuce.io/core/release/api/io/lettuce/core/codec/RedisCodec.html) to convert keys, so something like the following can be done to allow for applying custom RedisCodecs for key mappings (and ignoring the value mapper), without breaking existing code.

```java
    public static LettuceBasedProxyManagerBuilder<byte[]> builderFor(RedisClient redisClient) {
        return builderFor(redisClient.connect(ByteArrayCodec.INSTANCE));
    }

    public static <K> LettuceBasedProxyManagerBuilder<K> builderFor(RedisClient redisClient, RedisCodec<K, ?> keyCodec) {
        return builderFor(redisClient.connect(RedisCodec.of(keyCodec, ByteArrayCodec.INSTANCE)));
    }
```

With this one would be able to use `builderFor(redisClient, StringCodec.UTF8)` to create a `LettuceBasedProxyManager<String>` to allow for String keys

Happy to look into raising a PR for this if you are open to the idea

**Comment by vladimir-bukhtoyarov** [comment]
> Is this a feature on the radar for anytime soon?

It is not hard to implement, so I do not see objections to implement this featrue soon.

> For Lettuce at least, they provide [RedisCodec](https://lettuce.io/core/release/api/io/lettuce/core/codec/RedisCodec.html) to convert keys, so something like the following can be done to allow for applying custom RedisCodecs for key mappings

I have a little bit different architecture view about how does it should be implemented. The key point of each bucket4j integration - is utilizing as less third-party API as it possible. Following to this idea provides good surviving of the third-party API changes, usually Bucket4j does not need to be upgraded when third-part changed. So, I do not think that we should add new dependencies like RedisCodec from Lettuci and something similar from Jedis and Redisson. Instead it is enough to introduce new interface [Mapper](https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/distributed/serialization/Mapper.java) to the library and use Mapper directly inside LettuceBasedProxyManager(toBytes), JedisBasedProxyManager(toBytes) and RedissonBasedProxyManager(toString).

> Happy to look into raising a PR for this if you are open to the idea

Providing pull request will allow to bring closer the feature release. Use 8.3 branch as the base(and target).

**Reply by namhptran** [reply]
Thanks @vladimir-bukhtoyarov!

Given your current architectural plan, the Mapper interface you suggested looks good, though locking into mapping to String and `byte[]` would mean its not really usable for anything outside Redis

What's your opinion on a generic wrapper for ProxyManager which can be used as follows


```java

public interface ProxyManager<K> { 
  ...

    default <K2> ProxyManager<K2> withMapper(Function<? super K2, ? extends K> mapper) {
        return new ProxyManager<>() {
            @Override
            public RemoteBucketBuilder<K2> builder() {
                return ProxyManager.this.builder().withMapper(mapper);
            }

            @Override
            public Optional<BucketConfiguration> getProxyConfiguration(K2 key) {
                return ProxyManager.this.getProxyConfiguration(mapper.apply(key));
            }

            @Override
            public void removeProxy(K2 key) {
                ProxyManager.this.removeProxy(mapper.apply(key));
            }

            @Override
            public boolean isAsyncModeSupported() {
                return ProxyManager.this.isAsyncModeSupported();
            }

            @Override
            public AsyncProxyManager<K2> asAsync() throws UnsupportedOperationException {
                return ProxyManager.this.asAsync().withMapper(mapper);
            }
        };
    }
}
```

```java
ProxyManager<String> proxyManager = lettuceProxyManager.withMapper(Mappers.STRING_TO_BYTE)
```

**Reply by vladimir-bukhtoyarov** [reply]
@namhptran 

Proposed `ProxyManager.withMapper` is interesting idea, but it needs to be double thinked especially in scope of backends with strong schema. Currently, I prefer to solve problem locally inside `bucket4j-redis`.  I have created commit that passed all tests https://github.com/bucket4j/bucket4j/commit/88a66b7af1abfd1cec3868f0d7018acfd447ef83

Release process usually tooks a few hours, currently I have no such time window, but it looks that I will be able to release since 24 March

**Reply by namhptran** [reply]
Thanks for the quick progress with this @vladimir-bukhtoyarov. What are examples of these "backends with strong schema" that you mention might cause issues with this?

**Reply by vladimir-bukhtoyarov** [reply]
PostgreSQL, MySQL, DynamoDB. They obey user to specify schema of table, and type of primary key.

**Reply by namhptran** [reply]
Don't see how this would be a problem, as this example creates a proxy object around the original ProxyManager, so any underlying schema shouldn't affect it.

I have a proof of concept with working tests in this PR https://github.com/bucket4j/bucket4j/pull/360. Doesn't need to be included in the next release but maybe something for consideration if you like the idea.

**Reply by vladimir-bukhtoyarov** [reply]
I will look at weekend

**Comment by vladimir-bukhtoyarov** [comment]
Released with version 8.3.0
